extensions/default/src/DicomJSONDataSource/index.js

```diff
@@ -258,19 +262,19 @@ function createDicomJSONApi(dicomJsonConfig) {
 
       const { StudyInstanceUID, SeriesInstanceUID } = displaySet;
       const study = findStudies('StudyInstanceUID', StudyInstanceUID)[0];
-      const series = study.series.find(s => s.SeriesInstanceUID === SeriesInstanceUID);
-      let instances = displaySet.images;
-      if (series.instances.length > displaySet.images.length) {
-        instances = series.instances;
-      }
+      const series = study.series.find(s => s.SeriesInstanceUID === SeriesInstanceUID) || [];
 
-      instances.forEach(instance => {
+      displaySet.images.forEach(instance => {
```

Comment:
This algorithm is basically finding the right series.instances value, according to sop  instance uid and possibly frame number, so if you just create a map from  sop instance uid to the series instance that is relevant for that sop instance, then you can go through the loop without doing the series.instances filter query on every image, which reduces it from O(n^2) to O(n)
Something like
const seriesSopMap = new Map()
series.instances.forEach(   instance => addToSeriesSopMap(instance.sopInstanceUID, instance) ) 

Then just use seriesSopMap instead of sereis.instances.filter - that is, 
const images = seriesSopMap.get(instance.SOPInstanceUID)
const imageId = getImageId( {   instance: { ...(images.length > 0 ? images[i] : instance), frame: i, config: dicomJsonConfig })

Reply:
Updated!